File: 2013/treasure.py
#!/usr/bin/python

# Google Code Jam 2013 - Qualification Round - Problem D
#
# This code is able to solve the "small" input correctly in less than
# 0.1 seconds, and the "large" input (also correctly) in about 10
# seconds.
#
# Tobin Fricke 2013-04-13

import sys
from collections import Counter

debug = 0

# chests: set of unopened chests
# keys: multiset of keys that I have
# keys_req: dictionary chest-->key req
# keys_inside: dictionary chest-->set of keys inside

# Really, this only needs to be called once for a given problem
def impossible(chests, keys, key_req, keys_inside):
    keys_needed = Counter([key_req[c] for c in chests])
    keys_available = keys
    for c in chests:
        keys_available = keys_available + keys_inside[c]
    for k in keys_needed:
        if keys_available[k] < keys_needed[k]:
            if debug:
                print "Need %d of key %d but only have %d" % (keys_needed[k], k, keys_available[k])
            return 1  # impossible
    return 0 # not impossible

# see if we can succeed with "simultaneous unlocking"    
def possible(chests, keys, key_req, keys_inside):
    def openable(chest):
        return key_req[chest] in keys
    # Open everything we can without consuming keys:
    o = filter(openable, chests)
    while o:
        for c in o:
            keys = keys + keys_inside[c]
            chests = chests - {c}
        o = filter(openable, chests)
    return not chests

def solve(chests, keys, key_req, keys_inside):
    if not possible(chests, keys, key_req, keys_inside):
        return 0

    if not chests:
        return []

    for c in sorted(list(chests)):
        k = key_req[c]
        if k in keys:
            n = keys_inside[c]
            result = solve(chests - {c}, (keys - Counter({k})) + n, key_req, keys_inside)
            if isinstance(result, list):
                return [c] + result

    return 0

# T: number of test cases
T = int(raw_input())
if debug:
    print "=== %d test cases ===" % T

for t in range(0,T):
    if debug:
        print ""
    sys.stdout.write("Case #%d: " % (t+1))
    if debug:
        print ""
    # K: number of keys that I have
    # N: number of chests
    (K,N) = map(int, raw_input().split())

    # "This is followed by a line containing K integers, representing
    # the types of the keys that you start with."
    keys = Counter(map(int, raw_input().split()))
    if debug:
        print "I have these keys: " + str(keys)

    # After that, there will be N lines, each representing a single
    # chest. Each line will begin with integers Ti and Ki, indicating
    # the key type needed to open the chest and the number of keys
    # inside the chest. These two integers will be followed by Ki more
    # integers, indicating the types of the keys contained within the
    # chest.
    key_req = {};
    keys_inside = {};
    for n in range(1, N+1):
        numbers = map(int, raw_input().split())
        Ti = numbers.pop(0)
        Ki = numbers.pop(0)
        key_req[n] = Ti;
        keys_inside[n] = Counter({})
        for k in range(0, Ki):
            keys_inside[n] = keys_inside[n] + Counter([numbers.pop(0)])
        if debug:
            print ("Chest #%d needs key %d, contains %d keys: " % (n, Ti, Ki)) + str(keys_inside[n])
        if numbers:
            print "Error: unconsumed input: " + str(numbers)
    chests = set(range(1, N+1))
    sys.stdout.flush()
    if impossible(chests, keys, key_req, keys_inside):
        solution = 0
    else:
        solution = solve(frozenset(chests), keys, key_req, keys_inside)
    if isinstance(solution, list):
        print " ".join(map(str,solution))
    else:
        print "IMPOSSIBLE"
    sys.stdout.flush()

File: 2013/treasure.cc
#include <set>
#include <iostream>
#include <iterator>
#include <vector>

using namespace std;

bool debug = false;

bool global_constraint(set<int> chests, multiset<int> keys, int key_req[], multiset<int> keys_inside[]) {
  multiset<int> keys_available;
  multiset<int> keys_needed;

  keys_available.insert(keys.begin(), keys.end());

  for (set<int>::iterator c=chests.begin(); c != chests.end(); *c ++) {
    keys_available.insert(keys_inside[*c].begin(), keys_inside[*c].end());
    keys_needed.insert(key_req[*c]);
  }

  for (multiset<int>::iterator k=keys_needed.begin(); k!= keys_needed.end(); *k ++) {
    if (keys_needed.count(*k) > keys_available.count(*k)) {
      if (debug)
	cout << "Need " << keys_needed.count(*k) << " of key " << *k << " but only have " << keys_available.count(*k) << "." << endl;
      return false;
    }

  }
  return true;
}


bool still_possible(set<int> chests, multiset<int> keys, int key_req[], multiset<int> keys_inside[]) {
  bool flag = true;

  while (flag) {
    flag = false;
    
    for (set<int>::iterator c=chests.begin(); c != chests.end(); *c ++) {
      
      // check whether we can open the chest
      int key = key_req[*c];
      if (keys.count(key)==0)
	continue;
      
      // open it without destroying the key
      keys.insert(keys_inside[*c].begin(), keys_inside[*c].end());
      chests.erase(*c);
      flag = true; // there is still hope
      break;
    }
  }

  return chests.empty();
}

int solve(set<int> chests, multiset<int> keys, int key_req[], multiset<int> keys_inside[]) {

  int n_chests = chests.size();
  vector<int> solution;

  if (!global_constraint(chests, keys, key_req, keys_inside) ||
      !still_possible(chests, keys, key_req, keys_inside)) {
    cout << "IMPOSSIBLE" << endl;
    return 0;
  }

  while (!chests.empty()) {
    for (set<int>::iterator c=chests.begin(); c != chests.end(); *c ++) {
      // Skip this chest if we can't open it
      int chest = *c;
      int key = key_req[chest];
      if (keys.count(key)==0) {
	if (debug) 
	  cout << "Can't open chest " << chest + 1 << " because I don't have key " << key << endl;
	continue;
      }

      // Open the chest
      keys.erase(keys.find(key));
      set<int> new_chests = chests;
      new_chests.erase(chest);

      if (debug) {
	cout << "After removing chest " << chest+1 << ", new_chests contains: ";
	for (set<int>::iterator x=new_chests.begin(); x != new_chests.end(); *x ++) 
	  cout << *x + 1 << " ";
	cout << endl;
      }

      // insert keys_inside into keys
      for (multiset<int>::iterator k = keys_inside[chest].begin(); k != keys_inside[chest].end(); *k ++) {
	keys.insert(*k);
      }

      if (debug) {
	cout << "I have these keys: ";
	copy(keys.begin(), keys.end(),
	     ostream_iterator<int>(cout, " "));
	cout << endl;
      }

      // Test
      if (still_possible(new_chests, keys, key_req, keys_inside)) {
	chests = new_chests;    
	solution.push_back(chest + 1);
	break;
      }
      if (debug)
	cout << "Had to unopen chest " << chest + 1 << endl;
      // Rewind
      keys.insert(key);
      //chests.insert(*c);      
      // remove keys_inside from keys
      for (multiset<int>::iterator k = keys_inside[chest].begin(); k != keys_inside[chest].end(); *k ++) {
	keys.erase(keys.find(*k));
      }
    }
  }

  if (debug) 
    cout << "Solution: ";
  copy(solution.begin(), solution.end(),
       ostream_iterator<int>(cout, " "));
  cout << endl;
}


int main(int argc, char **argv) {
  int T;
  cin >> T;
  for (int t=0; t<T; t++) {
    cout << "Case #" << t+1 << ": ";
    if (debug) cout << endl;

    int K, N;
    cin >> K >> N;
    
    multiset<int> keys;
    for (int k=0; k<K; k++) {
      int temp;
      cin >> temp;
      keys.insert(temp);
    }
    if (debug) {
      cout << "I start with these keys: ";
      copy(keys.begin(), keys.end(),
	   ostream_iterator<int>(cout, " "));
      cout << endl;
    }

    int key_req[N];
    multiset<int> keys_inside[N];
    set<int> chests;

    for (int n=0; n<N; n++) {
      chests.insert(n);

      int Ti, Ki;
      cin >> Ti >> Ki;
      key_req[n] = Ti;
      for (int k=0; k<Ki; k++) {
	int temp;
	cin >> temp;
	keys_inside[n].insert(temp);
      }
      if (debug) {
	cout << "Chest " << n+1 << " needs key " << Ti << 
	  " and contains keys: ";
	copy(keys_inside[n].begin(), keys_inside[n].end(),
	     ostream_iterator<int>(cout, " "));
	cout << endl;
      }
    }
    
    solve(chests, keys, key_req, keys_inside);
  }

  return 0;
}
